core/shared/src/main/scala/zio/FiberRefs.scala

```diff
@@ -80,32 +107,36 @@ final class FiberRefs private (
     val parentFiberRefs = self.fiberRefLocals
     val childFiberRefs  = that.fiberRefLocals
 
-    val fiberRefLocals = childFiberRefs.foldLeft(parentFiberRefs) {
-      case (parentFiberRefs, (fiberRef, (childStack, childDepth))) =>
+    val fiberRefLocals0 = childFiberRefs.foldLeft(parentFiberRefs) {
+      case (parentFiberRefs, (fiberRef, Value(childStack, childDepth))) =>
         val ref        = fiberRef.asInstanceOf[FiberRef[Any]]
-        val childValue = childStack.head._2
+        val childValue = childStack.head.value
 
-        if (childStack.head._1 == fiberId) {
+        if (childStack.head.id == fiberId) {
           parentFiberRefs
         } else {
 
           parentFiberRefs
             .get(ref)
             .fold {
               if (childValue == ref.initial) parentFiberRefs
-              else parentFiberRefs.updated(ref, (::((fiberId, ref.join(ref.initial, childValue), 0), Nil), 1))
-            } { case (parentStack, parentDepth) =>
+              else
+                parentFiberRefs.updated(
+                  ref,
+                  Value(::(StackEntry(fiberId, ref.join(ref.initial, childValue), 0), List.empty), 1)
+                )
+            } { case Value(parentStack, parentDepth) =>
               @tailrec
               def findAncestor(
-                parentStack: List[(FiberId.Runtime, Any, Int)],
+                parentStack: List[StackEntry[?]],
                 parenthDepth: Int,
```

Comment:
There's something surprising in this function. 

This `parenthDepth` parameter seems to be misspelled, and so never used 🤔

The `parentDepth` variable, which is ("correctly spelled" and) used in the algorithm is always the one from the `.fold(...) { case Value(parentStack, parentDepth) => ... }` context, so is constant. 
This means that the recursive calls to `findAncestor`, when we `parentDepth - 1`, is always using the same value for `parentDepth - 1`. No matter how many recursive calls are made, the result of `parentDepth - 1` will always be the same 🤔

I'm not quite sure what this algo is doing nor what should be the correct value used in this substraction 🤔

Reply:
Ooof okay that's a bit scary. So [this is the PR](https://github.com/zio/zio/pull/8333) that added this, and it's been "wrong" since the beginning. Also, there were no tests added during that PR, so the new behaviour was not explicitly tested - so we can't rely on the test suite to tell us whether the current behaviour is correct.

I think that given the PR description, we should probably fix that typo and hope for the best? Or try and understand what the code is meant to be doing, but I'm really really struggling to follow the logic

Reply:
> but I'm really really struggling to follow the logic

Same 😕

Let's think about it. Right now, I don't know what's best. As said previously, I made a dedicated PR here #8748 to see what could be the impact on "fixing" this issue

Reply:
Oops, didn't see the linked PR in the comment. Thanks for opening that!

Reply:
Sadly, the tests still pass with the "fix" applied 😕

Reply:
@guizmaii since we're tracking this as a separate PR, I'll mark this as resolved in this PR